File: 2024/Day 12/Python/main.py
"""Advent of Code - 2024 - Day 12"""

def get_seed():
    """returns the first seed not in checked"""
    for y in range(h):
        for x in range(w):
            if (x, y) not in checked:
                return (x, y)
    print("error checked full")
    return (-1, -1)
with open(0, encoding="utf-8") as f:
    lines = f.read().splitlines()
w = len(lines[0])
h = len(lines)
checked: set[tuple[int, int]] = set()
price1: int = 0
price2: int = 0
while len(checked) < w * h:
    seed = get_seed()
    q = [seed]
    val = lines[seed[1]][seed[0]]
    region: set[tuple[int, int] ]= set()
    perimiter: int = 0
    corners: int = 0
    while q:
        n = q.pop()
        if n not in region:
            region.add(n)
            checked.add(n)
            west = (n[0]-1, n[1])
            left = 0 < n[0] and val == lines[west[1]][west[0]]
            if left:
                q.append(west)
            else:
                perimiter += 1
            east = (n[0]+1, n[1])
            right = n[0] < w-1 and val == lines[east[1]][east[0]]
            if right:
                q.append(east)
            else:
                perimiter += 1
            north = (n[0], n[1]-1)
            up = 0 < n[1] and val == lines[north[1]][north[0]]
            if up:
                q.append(north)
            else:
                perimiter += 1
            south = (n[0], n[1]+1)
            down = n[1] < h-1 and val == lines[south[1]][south[0]]
            if down:
                q.append(south)
            else:
                perimiter += 1

            neighbors = left + right + up + down
            if neighbors == 0:
                corners += 4
            elif neighbors == 1:
                corners += 2
            elif neighbors == 2:
                if up and left:
                    corners += 1
                    corners += lines[n[1]-1][n[0]-1] != val
                elif up and right:
                    corners += 1
                    corners += lines[n[1]-1][n[0]+1] != val
                elif down and right:
                    corners += 1
                    corners += lines[n[1]+1][n[0]+1] != val
                elif down and left:
                    corners += 1
                    corners += lines[n[1]+1][n[0]-1] != val
            elif neighbors == 3:
                if not up:
                    corners += lines[n[1]+1][n[0]+1] != val
                    corners += lines[n[1]+1][n[0]-1] != val
                elif not down:
                    corners += lines[n[1]-1][n[0]-1] != val
                    corners += lines[n[1]-1][n[0]+1] != val
                elif not left:
                    corners += lines[n[1]-1][n[0]+1] != val
                    corners += lines[n[1]+1][n[0]+1] != val
                elif not right:
                    corners += lines[n[1]-1][n[0]-1] != val
                    corners += lines[n[1]+1][n[0]-1] != val
            elif neighbors == 4:
                corners += lines[n[1]-1][n[0]-1] != val
                corners += lines[n[1]-1][n[0]+1] != val
                corners += lines[n[1]+1][n[0]+1] != val
                corners += lines[n[1]+1][n[0]-1] != val

    area: int = len(region)
    price1 += area * perimiter
    price2 += area * corners

print(price1)
print(price2)

File: 2024/Day 12/C++/main.cpp
#include <string>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <queue>
#include <algorithm>

std::pair<int, int> getSeed(int h, int w, std::unordered_set<int> &checked)
{
    for (int y = 0; y < h; y++)
    {
        for (int x = 0; x < w; x++)
        {
            if (checked.find(x + y * w) == checked.end())
            {
                return std::make_pair(x, y);
            }
        }
    }
    return std::make_pair(-1, -1);
}

int main(int argc, char *argv[])
{
    std::string line;
    std::vector<std::string> lines;
    while (std::getline(std::cin, line))
    {
        lines.push_back(line);
    }
    std::unordered_set<int> checked;
    int w = lines.size();
    int h = line.size();
    int price1 = 0;
    int price2 = 0;
    while (checked.size() < w * h)
    {
        std::pair<int, int> seed = getSeed(h, w, checked);
        std::queue<std::pair<int, int>> q;
        q.push(seed);
        char val = lines[seed.second][seed.first];
        std::vector<std::pair<int, int>> region;
        int perimiter = 0;
        int corners = 0;
        while (!q.empty())
        {
            std::pair<int, int> n = q.front();
            q.pop();
            if (std::find(region.begin(), region.end(), n) == region.end())
            {
                region.push_back(n);
                int x = n.first;
                int y = n.second;
                checked.insert(x + y * w);
                bool left = 0 < x && val == lines[y][x - 1];
                bool right = x < w - 1 && val == lines[y][x + 1];
                bool up = 0 < y && val == lines[y - 1][x];
                bool down = y < h - 1 && val == lines[y + 1][x];
                if (left)
                {
                    q.push(std::make_pair(x - 1, y));
                }
                else
                {
                    perimiter++;
                }
                if (right)
                {
                    q.push(std::make_pair(x + 1, y));
                }
                else
                {
                    perimiter++;
                }
                if (up)
                {
                    q.push(std::make_pair(x, y - 1));
                }
                else
                {
                    perimiter++;
                }
                if (down)
                {
                    q.push(std::make_pair(x, y + 1));
                }
                else
                {
                    perimiter++;
                }

                int neighbors = left + right + up + down;
                switch (neighbors)
                {
                case 0:
                    corners += 4;
                    break;
                case 1:
                    corners += 2;
                    break;
                case 2:
                    if (up && left)
                    {
                        corners++;
                        corners += (lines[y - 1][x - 1] != val);
                    }
                    else if (up && right)
                    {
                        corners++;
                        corners += (lines[y - 1][x + 1] != val);
                    }
                    else if (down && right)
                    {
                        corners++;
                        corners += (lines[y + 1][x + 1] != val);
                    }
                    else if (down && left)
                    {
                        corners++;
                        corners += (lines[y + 1][x - 1] != val);
                    }
                    break;
                case 3:
                    if (!up)
                    {
                        corners += (lines[y + 1][x + 1] != val);
                        corners += (lines[y + 1][x - 1] != val);
                    }
                    else if (!down)
                    {
                        corners += (lines[y - 1][x + 1] != val);
                        corners += (lines[y - 1][x - 1] != val);
                    }
                    else if (!left)
                    {
                        corners += (lines[y - 1][x + 1] != val);
                        corners += (lines[y + 1][x + 1] != val);
                    }
                    else if (!right)
                    {
                        corners += (lines[y - 1][x - 1] != val);
                        corners += (lines[y + 1][x - 1] != val);
                    }
                    break;
                case 4:
                    corners += (lines[y - 1][x - 1] != val);
                    corners += (lines[y - 1][x + 1] != val);
                    corners += (lines[y + 1][x + 1] != val);
                    corners += (lines[y + 1][x - 1] != val);
                    break;
                default:
                    break;
                }
            }
        }
        int area = region.size();
        price1 += area * perimiter;
        price2 += area * corners;
    }
    std::cout << price1 << "\n";
    std::cout << price2 << "\n";
}